toolchain/check/member_access.cpp

```diff
@@ -429,4 +428,81 @@ auto PerformCompoundMemberAccess(Context& context, Parse::NodeId node_id,
+    -> const llvm::APInt* {
+  const auto& index_val = context.ints().Get(index_inst.int_id);
+  if (index_val.uge(size)) {
+    CARBON_DIAGNOSTIC(
+        TupleIndexOutOfBounds, Error,
+        "Tuple element index `{0}` is past the end of type `{1}`.", TypedInt,
+        SemIR::TypeId);
+    context.emitter().Emit(node_id, TupleIndexOutOfBounds,
+                           {.type = index_inst.type_id, .value = index_val},
+                           operand_inst.type_id());
+    return nullptr;
+  }
+  return &index_val;
+}
+
+auto PerformTupleIndex(Context& context, Parse::NodeId node_id,
+                       SemIR::InstId tuple_inst_id, SemIR::InstId index_inst_id)
+    -> SemIR::InstId {
+  tuple_inst_id = ConvertToValueOrRefExpr(context, tuple_inst_id);
+  auto tuple_inst = context.insts().Get(tuple_inst_id);
+  auto tuple_type_id = tuple_inst.type_id();
+
+  CARBON_KIND_SWITCH(context.types().GetAsInst(tuple_type_id)) {
+    case CARBON_KIND(SemIR::TupleType tuple_type): {
+      SemIR::TypeId element_type_id = SemIR::TypeId::Error;
+      auto index_node_id = context.insts().GetLocId(index_inst_id);
+      index_inst_id = ConvertToValueOfType(
+          context, index_node_id, index_inst_id,
+          context.GetBuiltinType(SemIR::BuiltinInstKind::IntType));
+      auto index_const_id = context.constant_values().Get(index_inst_id);
+      if (index_const_id == SemIR::ConstantId::Error) {
+        index_inst_id = SemIR::InstId::BuiltinError;
+      } else if (!index_const_id.is_template()) {
+        // TODO: Decide what to do if the index is a symbolic constant.
+        CARBON_DIAGNOSTIC(TupleIndexNotConstant, Error,
+                          "Tuple index must be a constant.");
+        context.emitter().Emit(node_id, TupleIndexNotConstant);
+        index_inst_id = SemIR::InstId::BuiltinError;
+
+      } else {
+        auto index_literal = context.insts().GetAs<SemIR::IntLiteral>(
+            context.constant_values().GetInstId(index_const_id));
+        auto type_block = context.type_blocks().Get(tuple_type.elements_id);
+        if (const auto* index_val =
+                ValidateTupleIndex(context, node_id, tuple_inst, index_literal,
+                                   type_block.size())) {
+          element_type_id = type_block[index_val->getZExtValue()];
+        } else {
+          index_inst_id = SemIR::InstId::BuiltinError;
+        }
+      }
+
+      return context.AddInst<SemIR::TupleIndex>(node_id,
+                                                {.type_id = element_type_id,
+                                                 .tuple_id = tuple_inst_id,
+                                                 .index_id = index_inst_id});
+    }
+    default: {
+      if (tuple_type_id != SemIR::TypeId::Error) {
+        CARBON_DIAGNOSTIC(
+            TupleWrongIndexSyntax, Error,
+            "Type `{0}` does not support array indexing. Did you mean tuple.0?",
+            SemIR::TypeId);
```

Comment:
Should add a test that verifies that this diagnostic is being produced correctly. Also, isn't this the opposite case, where you are using tuple indexing on a non-tuple value?

Reply:
> Also, isn't this the opposite case, where you are using tuple indexing on a non-tuple value?

My bad, for some reason I was thinking of `tuple[0]`. Should that be handled btw?

Reply:
That would be great, but I wouldn't say it is required.